File: examples/python/Misc/Numeric.py
from __future__ import print_function

import math
import random

class Array:
    def __init__(self, matrix):
        self.matrix = matrix
        self.shape = shape(matrix)

    def __mul__(self, value):
        return Array(multiply(self.matrix, self.shape, value))

    def __rmul__(self, value):
        return Array(multiply(self.matrix, self.shape, value))

    def __sub__(self, value):
        return Array(subtract(self.matrix, self.shape, value))

    def __rsub__(self, value):
        return Array(subtract(self.matrix, self.shape, value))
        
    def __add__(self, value):
        if isinstance(value, (int, float)):
            return Array(add(self.matrix, self.shape, value))
        else:
            return Array(add_reduce_matrix([self.matrix, value]))
        
    def __radd__(self, value):
        if isinstance(value, (int, float)):
            return Array(add(self.matrix, self.shape, value))
        else:
            return Array(add_reduce_matrix([self.matrix, value]))
        
    def __pow__(self, value):
        return Array(pow(self.matrix, self.shape, value))

    def __gt__(self, value):
        return Array(gt(self.matrix, self.shape, value))

    def __lt__(self, value):
        return Array(lt(self.matrix, self.shape, value))

    def __len__(self):
        return len(self.matrix)

    def __getitem__(self, position):
        retval = self.matrix[position]
        print("getitem", position, self.matrix, retval)
        if isinstance(retval, (list, tuple)):
            return Array(retval)
        else:
            return retval

    def __setitem__(self, position, value):
        self.matrix[position] = value

    def tolist(self):
        return Array(self.matrix)

    def __repr__(self):
        return repr(self.matrix)


def ndim(n, *args):
    """
    Makes a multi-dimensional array of random floats. (Replaces RandomArray).
    """
    if not args: 
        return [random.random() for i in xrange(n)]
    A = [] 
    for i in range(n):
        A.append( ndim(*args) ) 
    return A 

def multiply(matrix, shape, value):
    if len(shape) == 1:
        return [m * value for m in matrix]
    else:
        return [multiply(matrix[i], shape[1:], value) for i in range(len(shape))]

def subtract(matrix, shape, value):
    if len(shape) == 1:
        return [m - value for m in matrix]
    else:
        return [subtract(matrix[i], shape[1:], value) for i in range(len(shape))]

def add(matrix, shape, value):
    print("add:", matrix, value)
    if len(shape) == 1:
        return [m + value for m in matrix]
    else:
        return [add(matrix[i], shape[1:], value) for i in range(len(shape))]

def pow(matrix, shape, value):
    if len(shape) == 1:
        return [m ** value for m in matrix]
    else:
        return [pow(matrix[i], shape[1:], value) for i in range(len(shape))]

def lt(matrix, shape, value):
    if len(shape) == 1:
        return [int(m < value) for m in matrix]
    else:
        return [lt(matrix[i], shape[1:], value) for i in range(len(shape))]

def gt(matrix, shape, value):
    if len(shape) == 1:
        return [int(m > value) for m in matrix]
    else:
        return [gt(matrix[i], shape[1:], value) for i in range(len(shape))]

def shape(matrix):
    if isinstance(matrix, (list, tuple)):
        return [len(matrix)] + shape(matrix[0])
    else:
        return []

def array(matrix, type='f'):
    """
    Might be an Array already? If so, copy the matrix.
    """
    return Array(matrix)

def zeros(size, type):
    if isinstance(size, (int, float)):
        return Array(ndim(int(size))) * 0.0
    else:
        return Array(ndim(*size)) * 0.0

def arange(size):
    """
    Range of values in array. Used in put.
    """
    return (0, size)

def put(arr, span, other):
    """
    other is a array or value
    """
    if isinstance(other, (list, tuple, Array)):
        for i in range(span[0], span[1]):
            arr[i] = other[i]
    else:
        for i in range(span[0], span[1]):
            arr[i] = other

def add_reduce(arr):
    return sum(arr)

def multiply_reduce(arr):
    retval = 0
    for v in arr:
        retval *= v
    return retval

def fabs(arr):
    """
    Absolute value of array.
    """
    if isinstance(arr, (float, int)):
        return abs(arr)
    else:
        return Array([fabs(v) for v in arr])

def argmax(arr):
    """
    Find maximum value, return position.
    """
    return arr.find(max(arr))

def matrixmultiply(a1, a2):
    print("numpy.dot(", a1, a2, ")")
    result = [a1[i] * a2[i] for i in range(len(a1))]
    result = add_reduce_matrix(result)
    print("return:", result)
    return result

def add_reduce_matrix(arr):
    results = arr[0]
    for i in range(1, len(arr)):
        for j in range(len(arr[i])):
            results[j] += arr[i][j]
    return results

def outerproduct(array1, array2):
    retval = []
    for item in array1:
        retval.append([item2 * item for item2 in array2])
    return retval

def exp(value):
    return math.exp(value)

def tanh(value):
    return math.tanh(value)

import os

print(os.path.dirname(__file__))

File: Calico.app/Contents/modules/Numeric/Numeric.cs
// C# Numeric replacement

// List
using System.Collections.Generic;
// IList
using System.Collections;

public static class Extensions
{
	public static T[] Slice<T> (this T[] source, int start)
	{
	        int end = source.Length;
		if (start == 0 && end == source.Length)
		    return source;
		int len = end - start;
		// Return new array.
		T[] res = new T[len];
		for (int i = 0; i < len; i++) {
		    res [i] = source [i + start];
		}
		return res;
	}

	public static T[] Slice<T> (this T[] source, int start, int end)
	{
		if (start == 0 && end == source.Length)
		    return source;
		// Handles negative ends.
		if (end < 0) {
		    end = source.Length + end;
		}
		int len = end - start;
		// Return new array.
		T[] res = new T[len];
		for (int i = 0; i < len; i++) {
		    res [i] = source [i + start];
		}
		return res;
	}

	public static object[] Slice (this IList source, int start, int end)
	{
		// Handles negative ends.
		if (end < 0) {
			end = source.Count + end;
		}
		int len = end - start;
		// Return new array.
		object[] res = new object[len];
		for (int i = 0; i < len; i++) {
			res [i] = source [i + start];
		}
		return res;
	}
}

public class Numeric {

    public delegate Function<double> Thunk ();

    public array ndim(int, params int args, Thunk thunk=null) {
	/*
	  Makes a multi-dimensional array of random floats. (Replaces
	  RandomArray).
	*/
	if (thunk == null) {
	    thunk = delegate { return random.random(); };
	}
	if (args.length == 0) {
	    return new List<double>([thunk.Invoke() for i in xrange(n)];
	}
		A = List<List<double>>();
	for (i in range(n)) {
	    A.append( ndim(args[0], args.Slice(1), thunk=thunk) ) ;
	}
	return A;
    }

    public class array {
	object array;

	public array (IList data, typecode='f') {
	    // Array([1, 2])
	    // Array([[1, 2], [3, 4]])
	    if (type(data) == array) {
		self.array = data[:];
	    } else if (type(data[0]) in [int, float, long, bool]) {
		// array.array of floats
		this.array = double [] (data);
	    } else {
		// list of Arrays
		self.array = map(array, data);
	    }
	}

    def __getitem__(self, item):
        return self.array[item]

    def __setitem__(self, item, value):
        self.array[item] = value

    def __len__(self):
        return len(self.array)

    def copy(self):
        return self.array[:]

    def __repr__(self):
        if type(self.array) is list:
            return str(map(lambda v: str(v.array.tolist()), self.array))
        else:
            return str(self.array.tolist())

    def __lt__(self, other):
        if type(self.array) is list:
            return array(map(lambda v: v < other, self.array))
        return array([f < other for f in self.array])

    def __gt__(self, other):
        if type(self.array) is list:
            return array(map(lambda v: v > other, self.array))
        return array([f > other for f in self.array])

    def __mul__(self, other):
        if type(other) in [int, float, long]:
            return array(map(lambda v: v * other, self.array))
        else: # array * [0, 1] maybe difference between list and array?
            return array(map(lambda a,b: a * b, self.array, other))

    def __sub__(self, other):
        if type(other) in [int, float, long]:
            return array(map(lambda v: v - other, self.array))
        else: # array - [0, 1]
            return array(map(lambda a,b: a - b, self.array, other))

    def __rsub__(self, other):
        if type(other) in [int, float, long]:
            return array(map(lambda v: other - v, self.array))
        else: # array - [0, 1]
            return array(map(lambda a,b: b - a, self.array, other))

    def __add__(self, other):
        if type(other) in [int, float, long]:
            return array(map(lambda v: v + other, self.array))
        else: # array + [0, 1]
            #print "add a", self.array
            #print "add b", other
            return array(map(lambda a,b: a + b, self.array, other))

    def __pow__(self, other):
        if type(other) in [int, float, long]:
            return array(map(lambda v: v ** other, self.array))
        else: # array ** [0, 1]
            return array(map(lambda a,b: a ** b, self.array, other))

    def __abs__(self):
        return array(map(lambda v: abs(v), self.array))

    def getShape(self):
        if type(self.array) == list:
            return (len(self.array), len(self.array[0]))
        else:
            return (len(self.array), )

    shape = property(getShape)
    __rmul__ = __mul__
    __rgt__ = __gt__
    __rlt__ = __lt__
    __radd__ = __add__
		}
fabs = abs
exp = math.exp

def put(toArray, arange, fromArray):
    for i in arange:
        if type(fromArray) in [int, float, long]:
            toArray[i] = fromArray
        else:
            toArray[i] = fromArray[i]

def arange(size):
    return range(size)

def zeros(dims, typecode='f'):
    if type(dims) == type(1):
        dims = (dims,)
    return array(ndim(*dims, thunk=lambda: 0.0))

public class add {
    public static object reduce(vector) {
        /*
	  Can be a vector or matrix. If data are bool, sum Trues.
        */
        if (type(vector) is list) { // matrix
            return array(map(add.reduce, vector));
        } else {
            return sum(vector); // Numeric_array, return scalar
	}
	
	public static array outerproduct(array a, array b) {
	    array result = zeros((len(a), len(b)));
	    for i in range(len(a)) {
		    for j in range(len(b)) {
			    result[i][j] = a[i] * b[j];
			}
		}
	    return result;
	}
    }
}

public class multiply {
    public static object reduce(vector) {
	/*
	  Can be a vector or matrix. If data are bool, sum Trues.
	*/
        if (type(vector) is list) { // matrix
            return array(map(multiply.reduce, vector));
	} else {
            return reduce(operator.mul, vector); // Numeric.array, return scalar
	}
    }
}

    public static array matrixmultiply(array a, array b) {
	//print "x a:", a
	//print "x b:", b
	// a = [[0, 1], [2, 3], [4, 5]], b = [0, 1]
    if type(b[0]) in [float, int, long] 
        retval = zeros(len(a))
        for i in range(len(a)):
            for j in range(len(a[0])):
                retval[i] = retval[i] + (a[i][j] * b[j])
    else:
        retval = zeros(len(b[0]))
        for i in range(len(a)):
            retval = retval + (a[i] * b[i])
    return retval;

		}